Parallel GH actions workflow for Nixpkgs eval

infinisil opened the pull request:
This PR primarily introduces a new GitHub Action workflow to effectively do a full Hydra evaluation for every PR, paving the way towards relieving ofborg of its evaluation duty. This is motivated by https://discourse.nixos.org/t/infrastructure-announcement-the-future-of-ofborg-your-help-needed/56025, see also https://github.com/NixOS/nixpkgs/issues/355847.

To make this as efficient as possible, 4 workflow jobs are spawned in parallel (one for each supported system, as originally experimented by @Mic92 in https://github.com/NixOS/nixpkgs/pull/352808), while each job makes full use of its 4 cores and 16GB of memory (based off amjoseph's idea in https://github.com/NixOS/nixpkgs/pull/269403 and initial work in https://github.com/NixOS/nixpkgs/pull/269356).

I've also implemented it in a way that limits maximum memory usage as Nixpkgs grows. CI will only take _longer_ to finish, instead of OOMing. Currently it takes about 5 minutes, check out [this test run](https://github.com/tweag/nixpkgs/actions/runs/11927279742?pr=99).

Note that ofborg does more than just evaluate Nixpkgs, but this PR can be used as a base to also implement output path comparisons to assign labels based on changed paths, to request appropriate maintainers, and do evaluation time comparisons.

Ping @philiptaron @Mic92 @Erethon @balsoft @JohnRTitor @a-kenji

## Things done
- [x] Tested a successful evaluation: https://github.com/tweag/nixpkgs/actions/runs/11927279742?pr=99
- [x] Tested a failed evaluation: https://github.com/tweag/nixpkgs/actions/runs/11927816023/job/33243842672?pr=99

---

This work is funded by [Tweag](https://tweag.io) and [Antithesis](https://antithesis.com/) :sparkles:

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
Parallel GH actions workflow for Nixpkgs eval

Motivated by ofborg struggling [1] and its evaluations taking too long,
inspired by Jörg's initial PR [2]
and Adam's previous attempt to parallelise Nixpkgs evaluation [3],
this PR contains initial work to relief ofborg from its evaluation duty
by using GitHub Actions to evaluate Nixpkgs.

For now this doesn't take care of all of what ofborg does, such as
requesting appropriate reviewers or labeling mass rebuilds, but this can
be follow-up work.

[1]: https://discourse.nixos.org/t/infrastructure-announcement-the-future-of-ofborg-your-help-needed/56025?u=infinisil
[2]: https://github.com/NixOS/nixpkgs/pull/352808
[3]: https://github.com/NixOS/nixpkgs/pull/269403

Co-Authored-By: Jörg Thalheim <joerg@thalheim.io>
Co-Authored-By: Adam Joseph <adam@westernsemico.com>

infinisil commented on .github/workflows/eval.yml:
```diff
@@ -0,0 +1,53 @@
+      - name: Enable swap
+        if: env.mergedSha
+        run: |
+          sudo fallocate -l 10G /swapfile
+          sudo chmod 600 /swapfile
+          sudo mkswap /swapfile
+          sudo swapon /swapfile
```
I can try it out, but are you sure? The [docs](https://docs.github.com/en/actions/using-github-hosted-runners/using-github-hosted-runners/about-github-hosted-runners#standard-github-hosted-runners-for-public-repositories) say that standard public runners have a 14GB SSD, while only the some [larger runners](https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners/about-larger-runners#specifications-for-general-larger-runners) seem to have a 75GB disk, and I can't find any other reference.

JohnRTitor replied:
Let's hope that they don't patch this out.

infinisil replied:
Actually with the parallel eval we don't even really need the swap. In the [test run](https://github.com/tweag/nixpkgs/actions/runs/11846539034/job/33014380928?pr=99) it was never needed (though just barely, but we can just increase the number of parallel chunks and therefore their size and memory usage).

Mic92 commented on .github/workflows/eval.yml:
```diff
@@ -0,0 +1,53 @@
+name: Eval
+
+on: pull_request_target
```
We can disable approval check in the github settings for our repository. The reason to not use it is the mentioned security issues and that we cannot test changes to the CI scripts this way.

infinisil replied:
The approval checks cannot be completely turned off unfortunately, the lowest setting is "Require approval for first-time contributors who are new to GitHub". I'm not sure what the setting was at, but last time we tried `pull_request` it was [quickly changed to `pull_request_target`](https://github.com/NixOS/nixpkgs/pull/254371).

Testing is a bit more annoying, but can be done by creating a separate PR, which is also more thorough, and I've been [doing](https://github.com/Infinisil-s-Test-Organization/nixpkgs/pulls?q=is%3Apr) [that](https://github.com/tweag/nixpkgs/pulls?q=is%3Apr) for a while. And if the PR that changes the workflow is not from a fork and instead using a Nixpkgs branch, we can also test it by making a separate PR against that branch.

And finally, `pull_request` wouldn't actually work here, because it would run without Nixpkgs secrets, which we need to request reviews (and just using `permissions: pull-requests: write` doesn't work, because it also needs permission to get org membership, because you can only request reviews from people that are in the org).

I'll take care to ensure there's a clear separation between potentially-malicious user input and secrets. This is indeed tricky as [seen before](https://github.com/NixOS/nixpkgs/pull/351446), but we have some good examples to learn from. I'd also appreciate @basti564 and 13x1's review here, and ideally also a link to the slides on GitHub Actions security you showed me at NixCon :)

Mic92 commented on .github/workflows/eval.yml:
```diff
@@ -0,0 +1,53 @@
+      - uses: cachix/install-nix-action@08dcb3a5e62fa31e2da3d490afc4176ef55ecd72 # v30
+        if: env.mergedSha
+
+      - name: Enable swap
+        if: env.mergedSha
+        run: |
+          sudo fallocate -l 10G /swapfile
+          sudo chmod 600 /swapfile
+          sudo mkswap /swapfile
+          sudo swapon /swapfile
```
If it is unused, we can just drop it for now. Especially given that we have another swap already enabled that can act as a canary:

```suggestion
```

JohnRTitor replied:
If allocating swap is not slowing down the gh action, I would recommend keeping it as a backup. You never know what happens.

infinisil replied:
I'll change it such that we have a constant chunk size, so that we effectively have an upper limit to memory usage

Mic92 replied:
@JohnRTitor there is already a swap file. So we would see if it is starting to swap before it swaps and can increase add this back as needed.

Mic92 commented:
I added several improvements to this script: https://github.com/tweag/nixpkgs/pull/100

TheGiraffe3 commented on .github/workflows/eval.yml:
```diff
@@ -0,0 +1,53 @@
+name: Eval
+
+on: pull_request_target
```
Nobody important here, but what about `push` as well?

Mic92 replied:
There is some code in this commit that wouldn't trigger on push.

But I think we will a have a different action later to compute the baseline store paths:

Here some code:

```
on:
  push:
    branches:
      - main
      - staging-*
      - release-*
      - python-updates
```

TheGiraffe3 replied:
I was thinking something like:
```suggestion
on: [push, pull_request_target]
```
So it'll run on any branch, not just the branches listed.

Mic92 replied:
I would rather not do that because than you are duplicating ci if someone opens a pull request from a branch in the repository. Also the github action you see here will not do anything if it runs on a branch because of the checks it does.

Mic92 commented:
I like this approach already. I want to incorporate this code in one form or the other in nixpkgs-review as it will make evaluating on normal laptops possible again.

JohnRTitor commented:
Please do! Currently evaluating uses too much memory and time!

Mic92 commented:
@GaetanLepage wants to look into integrating it into nixpkgs-review.

Mic92 commented:
@infinisil I would suggest to just address all comments for this pull request for now and maybe merge my pull request but not to extend the scope for the first pull request too much.
I would be good to first test how this action scales to the whole org before we proceed with more features.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/infrastructure-announcement-the-future-of-ofborg-your-help-needed/56025/37

JohnRTitor commented on ci/parallel.nix:
```diff
@@ -0,0 +1,63 @@
+}:
+
+let
+  attrPaths = builtins.fromJSON (builtins.readFile attrPathFile);
```
```suggestion
  attrPaths = lib.importJSON attrPathFile;
```

JohnRTitor commented on .github/workflows/eval.yml:
```diff
@@ -0,0 +1,45 @@
+            echo "Skipping the rest..."
+          fi
+          rm -rf base
+      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
```
I wonder why we are using commit hashes here instead of static tags?

infinisil replied:
https://github.com/NixOS/nixpkgs/pull/270552

JohnRTitor commented:
> It would be good to first test to see how this action scales to the whole org before we proceed with more features.

Agreed, due to slowness of Ofborg recently a lot of commiters are merging PRs without waiting for a successful eval which can take up to 13-15 hours. (I am [guilty of the act myself](https://github.com/NixOS/nixpkgs/pull/356363))

This fast but dangerous approach of review makes nixpkgs prone to a tree with a broken eval.

Having this basic CI check would help to detect issues beforehand, as we plan to phase out Ofborg in less than two months.

infinisil commented:
I did a rewrite and a rebase, check out https://github.com/tweag/nixpkgs/actions/runs/11888336167?pr=99 now, you can download the computed outpaths!

infinisil committed:
pkgs/top-level/release.nix: Don't include non-Hydra attributes with attrNamesOnly

The attrNamesOnly feature is used by
pkgs/top-level/release-attrpaths-superset.nix to return a superset of
all attributes that might be built by Hydra.

Before this change it would include all attribute paths to derivations that could
be found recursively, ignoring the recurseForDerivations and
recurseForRelease attributes that control Hydra's recursion.

This had the effect that it would end up with ~266000 attributes, most of which definitely won't be built by Hydra. We can remove those while staying true to the superset notion to end up with just ~97000, a reduction of ~63.6%! This also comes with an eval time reduction from 31.7s to 18.7s (-41.0%)!

As an example, all derivations in pypy310Packages don't get included
anymore, because it doesn't have a `.recurseForDerivations` set.

As a nice side effect, with `--arg enableWarnings false`, no more
warnings are printed, because things like
`checkpointBuildTools.mkCheckpointedBuild` (which is deprecated) isn't
being recursed to anymore.

infinisil commented:
This is done now imo! Pinging @Mic92 @Erethon @balsoft @JohnRTitor @a-kenji @azuwis @TheGiraffe3 for review and approval

TheGiraffe3 commented on ci/eval/README.md:
```diff
@@ -0,0 +1,19 @@
+# Nixpkgs CI evaluation
+
+The code in this directory is used by the [eval.yml](../../.github/workflows/eval.yml) GitHub Actions workflow to evaluate the majority of Nixpkgs for all PRs, effectively making sure that when the development branches are processed by Hydra, it doesn't encounter any evaluation failures.
```
```suggestion
The code in this directory is used by the [eval.yml](../../.github/workflows/eval.yml) GitHub Actions workflow to evaluate the majority of Nixpkgs for all PRs, effectively making sure that when the development branches are processed by Hydra, they don't encounter any evaluation failures.
```
Is "branches" or "Hydra" the subject?

infinisil replied:
Meant for it to be Hydra, how about avoiding the ambiguity by getting rid of the thing that could be ambiguous haha:
```suggestion
The code in this directory is used by the [eval.yml](../../.github/workflows/eval.yml) GitHub Actions workflow to evaluate the majority of Nixpkgs for all PRs, effectively making sure that when the development branches are processed by Hydra, no evaluation failures are encountered.
```

TheGiraffe3 replied:
Yep, your suggestion looks good to me.

TheGiraffe3 commented on ci/eval/README.md:
```diff
@@ -0,0 +1,19 @@
+
+- `--max-jobs`: The maximum number of derivations to run at the same time. Only each [supported system](../supportedSystems.nix) gets a separate derivation, so it doesn't make sense to set this higher than that number.
+- `--cores`: The number of cores to use for each job. Recommended to set this to the amount of cores on your system divided by `--max-jobs`.
+- `chunkSize`: The number of attributes that are evaluated simultaneously on a single core. Lowering this decreases memory usage at the cost of increased evaluation time. If this it too high, there won't be enough chunks to process them in parallel, also increasing evaluation time.
```
```suggestion
- `chunkSize`: The number of attributes that are evaluated simultaneously on a single core. Lowering this decreases memory usage at the cost of increased evaluation time. If this is too high, there won't be enough chunks to process them in parallel, and will also increase evaluation time.
```

TheGiraffe3 commented on ci/supportedSystems.nix:
```diff
@@ -0,0 +1,7 @@
+[
+  "aarch64-linux"
+  "aarch64-darwin"
+  #"i686-linux"  # !!!
```
Why is this commented out?

infinisil replied:
I kept it around because it was there already, but I just tracked it down via https://github.com/NixOS/ofborg/commit/db5293a4fb219d69f5734d87af573e1940396587 to https://github.com/NixOS/ofborg/commit/998adc8e2a7fddf9610d10a7657fb34c55492332, looks like it can be removed!

JohnRTitor replied:
When building graphics drivers, we need to build the corresponding 32 bit package as well. Since eval is not as "costly" anymore, can we restore this?

infinisil replied:
I'd say let's keep this PR focused on just mirroring what ofborg/Hydra do, especially because we need to see how GitHub Actions handles it for Nixpkgs volume.

JohnRTitor replied:
Let's remove it then.

TheGiraffe3 commented on pkgs/top-level/release-lib.nix:
```diff
@@ -163,19 +163,26 @@ let
+        f value
       else if value.recurseForDerivations or false || value.recurseForRelease or false then
-        packagePlatforms value
+        recursiveMapPackages f value
```
Apologies if I'm missing something, but does the "f" do anything?
```suggestion
        recursiveMapPackages value
```

infinisil replied:
It's a bit confusing, but `recursiveMapPackages` takes two arguments, because it's defined like `recursiveMapPackages f = mapAttrs (...)`, which is the same as `recursiveMapPackages f attrs = mapAttrs (...) attrs`. So here it's called recursively with the same `f` but a different `attrs`.

infinisil commented:
Follow-up fix: https://github.com/NixOS/nixpkgs/pull/357965

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/equinix-metal-cta-infra-short-update/56538/1

paparodeo commented:
I'm seeing ofborg eval failing on staging https://gist.github.com/GrahamcOfBorg/f50183ae86ce9f339b6b41cc33e14619 but the parallel evals are passing: https://github.com/NixOS/nixpkgs/pull/359106

FliegendeWurst commented:
staging will be fixed very soon. In this particular case the GH actions ran first, then staging broke, and finally ofborg got around to the PR.

infinisil commented:
WIP follow-up towards adding rebuild labels: https://github.com/NixOS/nixpkgs/pull/359704
